File: examples/reverse-proxy/dynamic_routes_proxy.py
"""
Dynamic route management — add and remove routes at runtime.

Demonstrates how to modify the proxy's route table while it is running.
This is useful when backends come and go (e.g. containers scaling up/down).

Usage:
    python dynamic_routes_proxy.py
"""

import asyncio

from wire import ReverseProxy


async def main():
    proxy = ReverseProxy(host="0.0.0.0", port=8080)

    # Start with a single backend
    proxy.add_route("/api", "http://localhost:3000")
    await proxy.start()
    print("Proxy started with /api -> http://localhost:3000")

    # Simulate a new service coming online after 5 seconds
    await asyncio.sleep(5)
    proxy.add_route("/metrics", "http://localhost:9090")
    print("Added /metrics -> http://localhost:9090")

    # Inspect current routes
    print("Current routes:", proxy.routes)

    # Simulate removing a route after another 10 seconds
    await asyncio.sleep(10)
    proxy.remove_route("/api")
    print("Removed /api route")
    print("Current routes:", proxy.routes)

    try:
        await asyncio.Future()
    except KeyboardInterrupt:
        pass
    finally:
        await proxy.stop()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/reverse-proxy/dynamic_routes_proxy.rs
/// Dynamic route management — add and remove routes at runtime.
///
/// Demonstrates how to modify the proxy's route table while it is running.
/// This is useful when backends come and go (e.g. containers scaling up/down).
///
/// Usage (from rust/wire-rs):
///     cargo run --example dynamic_routes_proxy

use wire_rs::proxy::ReverseProxy;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    env_logger::init();

    let mut proxy = ReverseProxy::new("0.0.0.0", 8080);

    // Start with a single backend
    proxy.add_route("/api", "http://localhost:3000").await;
    proxy.start().await?;
    println!("Proxy started with /api -> http://localhost:3000");

    // Simulate a new service coming online after 5 seconds
    tokio::time::sleep(std::time::Duration::from_secs(5)).await;
    proxy.add_route("/metrics", "http://localhost:9090").await;
    println!("Added /metrics -> http://localhost:9090");

    // Inspect current routes
    let routes = proxy.routes_snapshot().await;
    println!("Current routes: {:?}", routes);

    // Simulate removing a route after another 10 seconds
    tokio::time::sleep(std::time::Duration::from_secs(10)).await;
    proxy.remove_route("/api").await;
    println!("Removed /api route");

    let routes = proxy.routes_snapshot().await;
    println!("Current routes: {:?}", routes);

    tokio::signal::ctrl_c().await?;
    proxy.stop().await;
    Ok(())
}
